src/repos.cc

```diff
@@ -70,6 +70,27 @@ void safe_lexical_cast (Source s, Target &tr) {
   }
 }
 
+long long parse_priority(Zypper & zypper) {
+  long long prio = 0;
+  parsed_opts::const_iterator tmp1;
+  if ((tmp1 = zypper.cOpts().find("priority")) != zypper.cOpts().end())
+  {
+    //! \todo use some preset priorities (high, medium, low, ...)
+
+    string prio_str = *tmp1->second.begin();
+    safe_lexical_cast(prio_str, prio); // try to make an int out of the string
+    if (prio < 1)
+    {
+      zypper.out().error(boost::str(format(
+        _("Invalid priority '%s'. Use a positive integer number. The greater the number, the lower the priority."))
+        % prio_str));
+      zypper.setExitCode(ZYPPER_EXIT_ERR_INVALID_ARGS);
```

Comment:
I know,  the original code used ```long long```, but:
* repo priority should be ```unsigned parse_priority( Zypper & zypper )``` as this is what zypp:RepoInfo uses.
* __most important:__ if the cli arg is invalid (negative or no number) zypper must not continue! Add ```ZYPP_THROW(ExitRequestException("Invalid priority."));``` after ```setExitCode(...);```
* ```-p 0``` is (or should be) a legal priority value, setting the systems default priority: the value returned by ```static unsigned RepoInfo::defaultPriority()```.  
* it must be possible to distinguish between a valid ```-p arg``` and the absence of  ```-p``` on the command line. Thus ```0``` should exclusively be returned if no ```-p``` option is present (>0 for valid priority from CL; else throw).

Maybe something like:

```
unsigned parse_priority( Zypper & zypper )
{
  unsigned ret = 0U;
  parsed_opts::const_iterator cArg = zypper.cOpts().find("priority");
  if ( cArg == zypper.cOpts().end() )
    return ret;		// 0: no --priority arg

  int prio = -1;
  std::string prio_str = *cArg->second.begin();
  safe_lexical_cast(prio_str, prio); // try to make an int out of the string

  if ( prio < 0 )
  {
    zypper.out().error(boost::str(format(
      _("Invalid priority '%s'. Use a positive integer number. The greater the number, the lower the priority."))
      % prio_str));
    zypper.setExitCode(ZYPPER_EXIT_ERR_INVALID_ARGS);
    ZYPP_THROW(ExitRequestException("Invalid priority."));
  }

  ret = ( prio ? unsigned(prio) : RepoInfo::defaultPriority() );
  return ret;
}
```